### PR TITLE
Update etcd 2.2 references to use 3.0.x

### PR DESCRIPTION
**Pull request by timothysc**
This update an assortment of etcd 2.2.X references to 3.0.4 in the code base.  

/cc @hongchaodeng 

xref: https://github.com/kubernetes/kubernetes/issues/22448

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/29399)

### CODE REVIEW

**Comment by timothysc** [comment]
So when I try to grok the logs, I don't have visibility into where the cluster up is failing.

**Comment by hongchaodeng** [comment]
@timothysc 
There are some flags being lost in etcd v3. For example, the "-addr" is one of them. I have created an issue https://github.com/coreos/etcd/issues/6020 to address this and will resolve it ASAP. Thanks!

**Comment by xiang90** [comment]
@timothysc We need to use new flags suggested here: https://github.com/coreos/etcd/blob/release-2.0/etcdmain/config.go#L187-L193

This should be done when we upgraded etcd 0.4 -> 2.0. But we did not :(.

**Comment by hongchaodeng** [comment]
Sure.
@timothysc I will try to figure out the deprecated flags and send out another PR.

**Comment by justinsb** [comment]
Will this change cause the existing upgrade test suite to test the etcd upgrade?  (And we'll be operating in the same mode both after an upgrade and a direct install, right?)  Because until we have a fully functional upgrade procedure, we'll have to support both etcd2 and etcd3.  And if we are supporting both, we don't have to change the default :-)

**Comment by timothysc** [comment]
updating godep hitting: https://github.com/coreos/etcd/issues/6027

**Comment by hongchaodeng** [comment]
It looks like just the "hack/lib/etcd.sh" uses the deprecated flags.
Fix it in PR: #29458

**Comment by timothysc** [comment]
Working through the transitive dependencies from etcd 3 dep update causes rkt build issues now.  
/cc @kubernetes/sig-rktnetes 

```
# k8s.io/kubernetes/vendor/github.com/coreos/rkt/api/v1alpha
vendor/github.com/coreos/rkt/api/v1alpha/api.pb.go:1110: cannot use _PublicAPI_GetInfo_Handler (type func(interface {}, context.Context, func(interface {}) error) (interface {}, error)) as type grpc.methodHandler in field value
vendor/github.com/coreos/rkt/api/v1alpha/api.pb.go:1114: cannot use _PublicAPI_ListPods_Handler (type func(interface {}, context.Context, func(interface {}) error) (interface {}, error)) as type grpc.methodHandler in field value
vendor/github.com/coreos/rkt/api/v1alpha/api.pb.go:1118: cannot use _PublicAPI_InspectPod_Handler (type func(interface {}, context.Context, func(interface {}) error) (interface {}, error)) as type grpc.methodHandler in field value
vendor/github.com/coreos/rkt/api/v1alpha/api.pb.go:1122: cannot use _PublicAPI_ListImages_Handler (type func(interface {}, context.Context, func(interface {}) error) (interface {}, error)) as type grpc.methodHandler in field value
vendor/github.com/coreos/rkt/api/v1alpha/api.pb.go:1126: cannot use _PublicAPI_InspectImage_Handler (type func(interface {}, context.Context, func(interface {}) error) (interface {}, error)) as type grpc.methodHandler in field value
# k8s.io/kubernetes/pkg/storage/etcd3
pkg/storage/etcd3/compact.go:156: multiple-value client.KV.Compact() in single-value context
```

**Comment by wojtek-t** [comment]
@timothysc - actually, personally I think that we shouldn't do it that way. Etcd version shouldn't be hard-coded in the code. What we should do is to expose it as some "env variable" (probably in kube-up script) and propagate it correctly to manifests, scripts, etc.
We basically need to a way, to easily test both etcd2 and etcd3, since we can't stop testing etcd2 for some time at least.

@lavalamp - if you don't agree with that ^^

**Comment by timothysc** [comment]
@wojtek-t I may have time to clean up, but if not I'm sure we can do it in 2 parts.   Right now there is cruft, and I don't really know what is actually used and what should be removed.

**Comment by lavalamp** [comment]
This is intended to run the etcd3 binary in etcd2 mode?

**Comment by timothysc** [comment]
> This is intended to run the etcd3 binary in etcd2 mode?

Yes, I was planning to split up the PRs for final etcd3 + v3client enablement.

**Comment by onlyjob** [comment]
I wouldn't rush update of _etcd_ to v3 just yet: it uses "proto3" syntax files that can be built only with unreleased (beta) version of protobuf compiler...

**Comment by yifan-gu** [comment]
@timothysc @hongchaodeng FYI rkt has bumped grpc to the same commit as etcd 3.0.4 https://github.com/coreos/rkt/pull/3015 So you should be able to bump grpc in k8s now.

**Comment by hongchaodeng** [comment]
Yes. We are aware of that and have discussed to make the updates on upstream soon.

Thanks @yifan-gu @lucab and @xiang90 for coordinating!

**Comment by timothysc** [comment]
Yup, I'm catching up on my backlog I'll get to this soon.

**Comment by timothysc** [comment]
So resolved the transitive dep issues which was fun-zies.  I'm onto testing both etcd3 <> v2 and v3 clients.  I'll pastebin results for each.

**Comment by hongchaodeng** [comment]
```
FAILED   hack/make-rules/../../hack/verify-codecgen.sh
```
Did you change any generated code?

**Comment by timothysc** [comment]
I had to regen ```pkg/kubelet/api/v1alpha1/runtime/api.pb.go``` due to the protobuf dependency shift.

**Comment by timothysc** [comment]
Looks like I need to run a series of regen/update scripts

**Comment by timothysc** [comment]
node e2es are pending on https://github.com/kubernetes/kubernetes/pull/29267

File: cluster/images/hyperkube/static-pods/etcd.json [hunk 1]
```diff
@@ -10,7 +10,7 @@
     "containers": [
       {
         "name": "etcd",
-        "image": "gcr.io/google_containers/etcd-ARCH:2.2.5",
```
Comment (luxas) [inline cluster/images/hyperkube/static-pods/etcd.json]: This should be `-ARCH:3.0.3` and `3.0.3` should be pushed for all arches

Comment (timothysc) [inline cluster/images/hyperkube/static-pods/etcd.json]: Done.

**Comment by timothysc** [comment]
@lavalamp @wojtek-t - This PR is ready for review.  It updates the dependency shift to use etcd3 against the v2 api.  If possible, I would prefer not to conflate this PR with other build tooling change requests that should probably be done separately.  My plan is follow up with other PRs on the following: 

1. Update container to 3.0.4, and abstract etcd version to a global config to allow testing against either. 
2. Enable v3client + test cleanup.  
3. Documentation 

Thoughts? 
/cc @smarterclayton @ncdc @philips

**Comment by wojtek-t** [comment]
@timothysc - I definitely agree that this PR is big enough to not add anything else to it.

**Comment by lavalamp** [comment]
I will let @wojtek-t do the final review. I do want to know what we're supposed to do if we need to make a patch release for our etcd2 containers, start up scripts, etc?

File: build/build-image/cross/Dockerfile [hunk 2]
```diff
@@ -72,7 +72,7 @@ RUN mkdir $TMPDIR \
             github.com/jteeuwen/go-bindata/go-bindata
 
 # Download and symlink etcd. We need this for our integration tests.
-RUN export ETCD_VERSION=v2.2.1; \
+RUN export ETCD_VERSION=v3.0.3; \
```
Comment (wojtek-t) [inline build/build-image/cross/Dockerfile]: @timothysc - do we want to change it to 3.0.3 or to 3.0.4?

Comment (wojtek-t) [inline build/build-image/cross/Dockerfile]: Same question for all places.

File: pkg/kubelet/api/v1alpha1/runtime/api.pb.go [hunk 3]
```diff
@@ -230,17 +238,18 @@ type VersionResponse struct {
 	// The version of the kubelet runtime API.
 	Version *string `protobuf:"bytes,1,opt,name=version" json:"version,omitempty"`
 	// The name of the container runtime.
-	RuntimeName *string `protobuf:"bytes,2,opt,name=runtime_name" json:"runtime_name,omitempty"`
+	RuntimeName *string `protobuf:"bytes,2,opt,name=runtime_name,json=runtimeName" json:"runtime_name,omitempty"`
```
Comment (wojtek-t) [inline pkg/kubelet/api/v1alpha1/runtime/api.pb.go]: Changes to json field names looks suspicious to me, but IIUC, this is autogenerated code, right?

File: federation/apis/federation/types.go [hunk 4]
```diff
@@ -129,7 +129,7 @@ type FederatedReplicaSetPreferences struct {
 // Preferences regarding number of replicas assigned to a cluster replicaset within a federated replicaset.
 type ClusterReplicaSetPreferences struct {
 	// Minimum number of replicas that should be assigned to this Local ReplicaSet. 0 by default.
-	MinReplicas int64 `json:"maxReplicas,omitempty"`
+	MinReplicas int64 `json:"minReplicas,omitempty"`
```
Comment (wojtek-t) [inline federation/apis/federation/types.go]: good catch!

**Comment by wojtek-t** [comment]
>
I will let @wojtek-t do the final review. I do want to know what we're supposed to do if we need to make a patch release for our etcd2 containers, start up scripts, etc?

To summarize our yesterday discussion with @lavalamp 
- in 1.4 release we would like to change to etcd3 binary (no matter if we decide to recommend etcd v3 api or not) [fortunately it's possible to use etcd3 binary with etcd v2 api
- that said, if a bug exists that we need to fix, we should probably fix in 1.3 branch
- if a new bug exists with etcd3, we should fix it in branch

So I think that this PR shouldn't cause any problems in the future.


@timothysc - I added two comments

File: build/build-image/cross/Dockerfile (hunk 2 above)
Comment (timothysc) [inline build/build-image/cross/Dockerfile]: Ideally yes, but we'd need the container pushed to google repos, I can't make that happen through the PR process.

Comment (wojtek-t) [inline build/build-image/cross/Dockerfile]: Got it - let's leave as is and we can change that in a followup PR.

File: pkg/kubelet/api/v1alpha1/runtime/api.pb.go (hunk 3 above)
Comment (timothysc) [inline pkg/kubelet/api/v1alpha1/runtime/api.pb.go]: This is all generated, it should probably be ripped out of this commit as it's overwritten in the other.

**Comment by timothysc** [comment]
@wojtek-t any issues?

**Comment by wojtek-t** [comment]
I'm fine with this PR - lgtm

**Pull request merged**